### ISSUE TITLE
Spurious "Condition will always evaluate to false" with inspect.Parameter

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#2708]
On this code:
```python
import inspect
from typing import Iterable


def render_parameters(param: inspect.Parameter) -> Iterable[str]:
    kind = param.kind

    if kind == inspect.Parameter.POSITIONAL_ONLY:
        pass

    if kind == inspect.Parameter.VAR_POSITIONAL:
        pass
    elif kind == inspect.Parameter.KEYWORD_ONLY:
        yield "*"
```
Pyright produces
`Condition will always evaluate to False since the types "Literal[_ParameterKind.POSITIONAL_ONLY]" and "Literal[_ParameterKind.VAR_POSITIONAL]" have no overlap` on both of the last comparisons.

I couldn't reproduce this with an enum defined inline, so I suspect it has something to do with the definition of the ParameterKind enum in typeshed, which doesn't have values for the enum members (https://github.com/python/typeshed/blob/master/stdlib/inspect.pyi#L200).

**Comment by erictraut** [comment]
Yeah, there are two bugs here. One is in pyright, which isn't correctly dealing with an enum that has no enumerated members.

The second is in typeshed, which is incorrectly defining the enum. Enumerated members are supposed to be annotated using an assignment operator. Attributes within an enum class that use a type annotation are assumed to be non-enum attributes. For example, `description: str` (which is also declared in the `_ParameterKind` class) is presumably not part of the enumeration but rather an attribute on the class.

The typeshed definition should be changed to:
```python
class _ParameterKind(enum.IntEnum):
    POSITIONAL_ONLY = ...
    POSITIONAL_OR_KEYWORD = ...
    VAR_POSITIONAL = ...
    KEYWORD_ONLY = ...
    VAR_KEYWORD = ...

    if sys.version_info >= (3, 8):
        description: str
```

**Comment by JelleZijlstra** [comment]
Thanks! Changing typeshed may be a big project because defining enums this way is the de facto convention in typeshed. I actually introduced it myself in python/typeshed#3168. If we have to change the convention, we should open an issue in typeshed and then document the decision in typeshed's CONTRIBUTING.

Your convention has the disadvantage that it doesn't tell the typechecker what the `.value` of the enum's members is. An alternative could be to put in the actual runtime values, like `POSITIONAL_ONLY = 1`.

**Comment by erictraut** [comment]
Actually, I stand corrected. I forgot that I had added different handling for enum definitions when they are found in type stub files. This was specifically to accommodate typeshed's convention. I'll need to investigate further on this issue then.

We should still have the discussion about enum typing conventions in typeshed because I don't think the current convention can distinguish between enumerated members and non-enumerated attributes on the class. For example, how would you expect the `_ParameterKind .description` field to be interpreted by a type checker.

**Comment by JelleZijlstra** [comment]
The convention I adopted in the linked PR is to use `@property` for attributes like the `description` field. As it happens, `_ParameterKind.description` really is a property at runtime, so I'll send a PR to typeshed to fix it. I'm not sure that solution will always work though—I know you have been pushing to avoid using `@property` for things that may not use the property decorator in reality, like readonly attributes on protocols.

**Comment by erictraut** [comment]
OK, I found a separate bug in pyright that explains the behavior you're seeing with `_ParameterKind`. It would affect type narrowing for any enum declared in a stub, so it's a pretty bad bug. The root cause is due to the alternate behavior that I added for type stubs, which was violating an assumption in the type guard logic. This bug will be fixed in the next version.

**Comment by erictraut** [comment]
This is included in pyright 1.1.197, which I just published. It will also be included in the next release of pylance.
